### PR TITLE
make cdot reader left associative

### PR DESCRIPTION
**Pull request by AlexKnauth**
So that with the `read-cdot` option turned on,
```racket
X.Y.Z
```
Reads equal to 
```racket
(#%dot (#%dot X Y) Z)
```
This is what most language implementers will expect when they are writing their `#%dot` macros.

Resolves https://github.com/racket/racket/issues/1443.

### CODE REVIEW

**Comment by LeifAndersen** [comment]
Yes, very much this.

File: pkgs/racket-doc/scribblings/reference/reader.scrbl [hunk 1]
```diff
@@ -956,12 +956,16 @@ numbers are followed by a @litchar{.} intended to be read as a C-style
...
 
 Finally, after reading any value, @racket[_x], the reader will seek
-over whitespace until it reaches a non-whitespace character. If the
-character is not @litchar{.}, then the value, @racket[_x], is returned
-as usual. If the character is @litchar{.}, then another value,
-@racket[_y], is read and the result @racket[(list '#%dot _x _y)] is
-returned. In @racket[read-syntax] mode, the @racket['#%dot] symbol has
-the source location information of the @litchar{.} character and the
+through whitespace and look for zero or more sequences of a
+@litchar{.} followed by another value @racket[_y], to put each
```
Comment (bennn) [inline pkgs/racket-doc/scribblings/reference/reader.scrbl]: While we're here, I think "value" should be "datum", re Jay's comment on the issue.

File: pkgs/racket-doc/scribblings/reference/reader.scrbl [hunk 2]
```diff
@@ -956,12 +956,16 @@ numbers are followed by a @litchar{.} intended to be read as a C-style
...
+@racket[_y] together with @racket[_x] in a @racket[#%dot] form.
+@racket[_x.y] will be read as @racket[(#%dot _x _y)].
+
+It does this in a left-associative way so that @racket[_x.y.z]
+will be read as @racket[(#%dot (#%dot _x _y) _z)].
```
Comment (bennn) [inline pkgs/racket-doc/scribblings/reference/reader.scrbl]: Suggestion: replace this sentence with "If `y` contains a `.`, the reader will group `.`-separated datums from left-to-right. For example, `x.y.z` reads as `(list (list #%dot x y) z)`.

File: pkgs/racket-doc/scribblings/reference/reader.scrbl (hunk 1 above)
Comment (AlexKnauth) [inline pkgs/racket-doc/scribblings/reference/reader.scrbl]: Yes, definitely.

File: pkgs/racket-doc/scribblings/reference/reader.scrbl (hunk 2 above)
Comment (jeapostrophe) [inline pkgs/racket-doc/scribblings/reference/reader.scrbl]: +1

**Comment by jeapostrophe** [comment]
You should update `read-cdot`'s existing clients because this is a backwards-incompatible change.

File: pkgs/racket-doc/scribblings/reference/reader.scrbl (hunk 2 above)
Comment (AlexKnauth) [inline pkgs/racket-doc/scribblings/reference/reader.scrbl]: It's not really that `y` contains a `.`, it's that there's another `.` after `y`. So would "If `y` has another `.` after it, the reader will group `.`-separated datums from left-to-right" be better?

**Comment by jeapostrophe** [comment]
Once you are happy with the docs, it is fine to launch by me

**Comment by LeifAndersen** [comment]
Hey @jeapostrophe , since we're breaking backwards compatibility, do you have any opinions on changing the name to something other than cdot? As cdot doesn't make sense as a name unless you know much about c-style languages. Also, cdot means something completely different in say latex, which might confuse things.

**Comment by bennn** [comment]
@LeifAndersen how about `read-#%dot` ?

**Comment by AlexKnauth** [comment]
That can be a separate change. I would rather do one thing at a time

**Comment by bennn** [comment]
@AlexKnauth ok. Are you planning to update the existing clients in this PR?

**Comment by AlexKnauth** [comment]
@bennn Yes, there's a pull request ready here: https://github.com/jeapostrophe/remix/pull/1

**Comment by LeifAndersen** [comment]
@bennn I'm fairly ambivalent on read-#%dot.  I'm not aware of anything else in our standard library that follows a convention like that though.

**Comment by LeifAndersen** [comment]
Some suggestions (with help from Ben) might be: `read-dot-notation`, also `read-dot-operator` would be nice too, although I guess that very much goes against Jay saying dot is not an operator, so maybe not the second one.

**Comment by AlexKnauth** [comment]
Since dots already mean 3 different things in racket without this, those would all be confusing. The name `cdot` at least differentiates it from the more lisp-based uses of dots.

**Comment by LeifAndersen** [comment]
Ya, that's fair.

**Comment by jeapostrophe** [comment]
I don't think we should change the name and I don't think we should stress about the backwards-incompat given that there is one use and no one uses it. I'm good to commit.

**Comment by AlexKnauth** [comment]
Okay. Can you look over the doc changes and see if there's anything else that should be changed?

**Comment by LeifAndersen** [comment]
Ya, I'm inclined to agree with @jeapostrophe here. As sad as it is, I can't think of anything better then cdot, oh well. :( We should probably not change the name, as its not really that important.

**Comment by jeapostrophe** [comment]
I think everything is good and I can merge it.

**Pull request merged**